cpp/recognition/cograph_recognition/Recognition.cpp

```diff
@@ -0,0 +1,331 @@
+                t->md = 0;
+                t = t->parent->parent;
+            }
+            w = u;
+        }
+        return {w, CorneilStewartPerlCographRecognition::State::COGRAPH};
+    }
+
+
+    std::vector<CoNode *> GetWereMarked(CoNode *u) {
+        auto x = u->head_of_list_of_children;
+        std::vector<CoNode *> a;
+        while (x != nullptr && x->marked == Marked::MARKED_AND_UNMARKED) {
+            a.push_back(x);
+            x = x->next;
+        }
+        return a;
+    }
+
+
+    CoNode *GetLastFromChildren(CoNode *u) {
+        auto x = u->head_of_list_of_children;
+        while (x != nullptr && x->marked == Marked::MARKED_AND_UNMARKED) {
+            x = x->next;
+        }
+        return x;
+    }
+
+    void InsertXToCoTree(CoNode *u, CoNode *x) {
+        std::vector<CoNode *> a;
+        int u_number = u->number;
+        a = GetWereMarked(u);
+        if ((a.size() == 1 && u_number == 0) || (u->d - a.size() == 1 && u_number == 1)) {
+            CoNode *w = a[0];
+            if (u_number == 1) {
+                w = GetLastFromChildren(u);
+            }
+            if (w->type == Type::VERTEX) {
+                auto *y = new CoNode(Type::ZERO_ONE, u_number ^ 1);
+                T->Add(y);
+                if (u_number == 0) {
+                    u->RemoveWereMarked();
+                } else {
+                    u->RemoveWereNotMarked();
+                }
+                u->AddChild(y);
+                y->AddChild(x);
+                y->AddChild(w);
+            } else {
+                w->AddChild(x);
+            }
+        } else {
+            auto vec = u->RemoveWereMarked();
+            auto *y = new CoNode(Type::ZERO_ONE, u_number);
+            T->Add(y);
+            for (auto v: vec) {
+                y->AddChild(v);
+            }
+            if (u_number == 1) {
+                auto nxt = u->next;
+                auto prv = u->prev;
+                if (prv != nullptr) {
+                    prv->next = y;
+                }
+                if (nxt != nullptr) {
+                    nxt->prev = y;
+                }
+                y->prev = prv;
+                y->next = nxt;
+                if (prv == nullptr && u->parent != nullptr) {
+                    u->parent->head_of_list_of_children = y;
+                }
+                if (u->parent != nullptr) {
+                    y->parent = u->parent;
+                } else {
+                    T->root = y;
+                }
+                auto *z = new CoNode(Type::ZERO_ONE, 0);
+                T->Add(z);
+                y->AddChild(z);
+                z->AddChild(x);
+                z->AddChild(u);
+            } else {
+                auto *z = new CoNode(Type::ZERO_ONE, 1);
+                T->Add(z);
+                u->AddChild(z);
+                z->AddChild(x);
+                z->AddChild(y);
+            }
+        }
+    }
+
+
+    CorneilStewartPerlCographRecognition::State
```

Comment:
If you find the return types too long, consider using `auto` instead - similarly e.g. for `FindLowest()`

Reply:
left